modules/adlooxRtdProvider.js

```diff
@@ -213,193 +99,119 @@ function init(config, userConsent) {
-  ];
-
-  if (!adUnits.length) {
-    logWarn(MODULE, 'no suitable adUnits (missing pbadslot?)');
-  }
-  const atfQueue = [];
-  adUnits.map((adUnit, i) => {
-    const ref = [ adUnit.gpid ];
-    if (!config.params.slotinpath) ref.push(adUnit.unit.code);
-    args.push(['s', ref.join('\t')]);
-
-    semaphoreInc(1);
-    atfQueue.push(function() {
-      atf(adUnit.unit, function(x) {
-        let viewable = document.visibilityState === undefined || document.visibilityState == 'visible';
-        try { viewable = viewable && top.document.hasFocus() } catch (_) {}
-        logInfo(MODULE, `atf code=${adUnit.unit.code} has area=${x}, viewable=${viewable}`);
-        const atfList = []; atfList[i] = { atf: parseInt(x * 100) };
-        response = mergeDeep(response, { _: atfList });
-        semaphoreInc(-1);
-      });
-    });
-  });
-  function atfCb() {
-    atfQueue.forEach(x => x());
-  }
-  if (document.readyState == 'loading') {
-    document.addEventListener('DOMContentLoaded', atfCb, false);
-  } else {
-    atfCb();
-  }
-
-  analyticsCommand(COMMAND.URL, {
-    url: (config.params.api_origin || API_ORIGIN) + '/q?',
-    args: args
-  }, function(url) {
-    ajax(url, {
-      success: function(responseText, q) {
-        try {
-          if (q.getResponseHeader('content-type') == 'application/json') {
-            response = mergeDeep(response, JSON.parse(responseText));
-          } else {
-            throw false;
-          }
-        } catch (_) {
-          logError(MODULE, 'unexpected response');
-        }
-        semaphoreInc(-1);
-      },
-      error: function(statusText, q) {
-        logError(MODULE, 'request failed');
-        semaphoreInc(-1);
-      }
-    });
-  });
+  );
 }
 
-function getTargetingData(adUnitArray, config, userConsent, auction) {
-  function targetingNormalise(v) {
+function getTargetingData(adUnitArray, config, userConsent) {
+  function val(v) {
     if (isArray(v) && v.length == 0) return undefined;
     if (isBoolean(v)) v = ~~v;
     if (!v) return undefined; // empty string and zero
     return v;
   }
 
-  const ortb2 = auction.getFPD().global || {};
-  const dataSite = deepAccess(ortb2, `site.ext.data.${MODULE_NAME}_rtd`) || {};
-  if (!dataSite.ok) return {};
+  const targeting = {};
 
-  const dataUser = deepAccess(ortb2, `user.ext.data.${MODULE_NAME}_rtd`) || {};
-  return getGlobal().adUnits.filter(adUnit => includes(adUnitArray, adUnit.code)).reduce((a, adUnit) => {
-    a[adUnit.code] = {};
+  const ortb2base = {};
+  _each(config._fpd, function(v0, k) {
+    const v = val(v0);
+    if (v) ortb2base[`${ADSERVER_TARGETING_PREFIX}_${k}`] = v;
+  });
 
-    _each(dataSite, (v0, k) => {
-      if (includes(SEGMENT_HISTORIC_VALUES, k)) return; // ignore site average viewability
-      const v = targetingNormalise(v0);
-      if (v) a[adUnit.code][ADSERVER_TARGETING_PREFIX + k] = v;
-    });
+  const adUnits = getGlobal().adUnits.filter(unit => adUnitArray.includes(unit.code));
```

Comment:
this should be `const adUnits = auction.adUnits`. There's again a race condition with using the global adUnits, since a publisher could do:

```
pbjs.addAdUnits(...)
pbjs.requestBids(...)
pbjs.removeAdUnit(...)
```

And by the time the auction ends and this code runs, you are working on the wrong ad units.

In fact I think it's likely a bunch of other adapters are doing this, since the original API gave them no other option. I'll take a look and probably open a separate issue.

Reply:
I mean that by the time this code runs (at the end of the auction), `getGlobal().adUnits` might be completely different from what it was when the auction started - and your filter would suffer from the crap in, crap out problem. In fact, if the publisher does a completely legitimate

```
pbjs.requestBids({adUnits: [...]})
```

then `getGlobal().adUnits` would be empty the whole time. I think the RTD module API was defined a time when neither the `adUnits` argument nor concurrent auctions were supported, so it's a good thing we finally spotted this.

Reply:
The design of our module is not sensitive to if `getGlobal().adUnits` changes between the call to `getBidRequestData` and `getTargetingData` or before the call to `getBidRequestData` runs to completion.

For adUnit segments, within `getBidRequestData` we operate on our own array of `adUnit` objects cloned from the original list. If we set the FPD on *objects* that are no longer part of the original array (ie. deleted), it does not matter as the result will just be effectively discarded.

The site/user/device segments are queried once and always the same (per-referrer/device) and as such only those go into `config._fpd`. Poking our segment endpoint will return stable site/user/device results for the same page/device/user.

By the time we get to `getTargetingData`, we grab all adUnits and only do something when `AdSlot.ortb2Imp.ext.data.adloox_rtd` is present...otherwise we NOOP. So no harm if those slots we previously queried have now been deleted.

This is why I stash and try to extract the data from the adUnits themselves so we avoid storing state within our own module; which other RTD modules look to do.

The problem our RTD module does have though is it assumes `getRefererInfo` is a constant, which it seems not to be. This will need fixing, I'll do this whilst putting in back the use of `auction` (thanks for pointing how to access that safely!).

Reply:
> So no harm if those slots we previously queried have now been deleted.

The issue is that from the pub's POV they have not deleted them; they have used them in an auction, presumably expecting the adloox targeting logic to work on them, and then moved them out of the way to start another auction (before the targeting logic has had a chance to run). So no harm, but also no functionality :) 

All of this is nitpicking, since refactoring as we discussed will fix the problem.

Reply:
@dgirardi okay, thank you for the help, pointers and advice here. I have reworked the module to use the `auction` object and try to steer clear of `getGlobal().adUnits`, let me know if things look better